tests/tracer/test_trace_utils.py

```diff
@@ -580,6 +593,95 @@ def test_set_http_meta_headers_ip(
             mock_store_headers.assert_called()
 
 
+@pytest.mark.parametrize(
+    "headers_dict,peer_ip,expected",
+    [
+        # Public IP in headers should be selected
+        (
+            {"x-forwarded-for": "8.8.8.8"},
+            "127.0.0.1",
+            "8.8.8.8",
+        ),
+        # Public IP in headers only should be selected
+        (
+            {"x-forwarded-for": "8.8.8.8"},
+            "",
+            "8.8.8.8",
+        ),
+        # Public peer IP, get preferred over loopback IP in headers
+        (
+            {"x-forwarded-for": "127.0.0.1"},
+            "8.8.8.8",
+            "8.8.8.8",
+        ),
+        # Public peer IP, get preferred over private IP in headers
+        (
+            {"x-forwarded-for": "192.168.1.1"},
+            "8.8.8.8",
+            "8.8.8.8",
+        ),
+        # Public IP on both, headers selected
+        (
+            {"x-forwarded-for": "8.8.8.8"},
+            "8.8.4.4",
+            "8.8.8.8",
+        ),
+        # Empty header IP, peer should be selected
+        (
+            {"x-forwarded-for": ""},
+            "8.8.4.4",
+            "8.8.4.4",
+        ),
+        # Empty header IP, peer should be selected even if loopback
+        (
+            {"x-forwarded-for": ""},
+            "127.0.0.1",
+            "127.0.0.1",
+        ),
+        # Empty headers, peer should be selected even if loopback
+        (
+            {},
+            "127.0.0.1",
+            "127.0.0.1",
+        ),
+        # None headers, peer should be selected even if loopback
+        (
+            None,
+            "127.0.0.1",
+            "127.0.0.1",
+        ),
+        # None everything, empty IP returned
+        (
+            None,
+            None,
+            "",
+        ),
+        # None headers, invalid peer IP, empty IP returned
+        (
+            None,
+            "invalid",
+            "",
+        ),
+        # Both invalid, empty IP returned
+        (
+            {"x-forwarded-for": "invalid"},
+            "invalid",
+            "",
+        ),
+        # Invalid IP in headers, peer ip should be selected even if loopback
```

Comment:
nice comments :clap:

Reply:
Yeah, wrote them for me mostly since I had to mentally parse the fixtures from the other test above :)